### ISSUE TITLE
Incorrect type checking of protocol with independent TypeVars

### ISSUE THREAD

**Issue by Elchinchel** [microsoft/pyright#6213]
I wrote protocol class, which has function with default value (same as dict.get) and after implementing protocol in other library discover this behaviour.

**Code**
With following code, pyright show error on last line:

```python
from typing import TypeVar, Protocol, Generic


# there is same error if _Data1 and _Data2 are invariant
_Data1 = TypeVar('_Data1', covariant=True)
_Default1 = TypeVar('_Default1')

class KeyStorage(Protocol[_Data1]):
    def get(self, __key: str, __default: _Default1) -> '_Data1 | _Default1': ...


_Data2 = TypeVar('_Data2', covariant=True)
_Default2 = TypeVar('_Default2')

class RealKeyStorage(Generic[_Data2]):
    def get(self, key: str, default: _Default2) -> '_Data2 | _Default2':
        raise NotImplementedError


def foo(storage: KeyStorage[int]):
    reveal_type(storage.get('foo', None))  # Type of "storage.get('foo', None)" is "int | None"


foo_storage: RealKeyStorage[str] = RealKeyStorage()

foo(foo_storage)
# Argument of type "RealKeyStorage[int]" cannot be assigned to parameter "storage" of type "KeyStorage[int]" in function "foo"
#     "RealKeyStorage[int]" is incompatible with protocol "KeyStorage[int]"
#       Type parameter "_Data1@KeyStorage" is covariant, but "int | _Default2@get" is not a subtype of "int"
#         Type "int | object*" cannot be assigned to type "int"
#           "object*" is incompatible with "int"
```

**Versions**
Noticed this behaviour in VSCode with **Pylance v2023.10.30**, tried pre-release, got same result.
**pyright 1.1.332** report same error, **mypy 1.6.1** report no errors.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.333, which I just published. It will also be included in a future release of pylance.
